### PR TITLE
BUG: Localised copy support for ClassName changes.

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
# Localised copy support for ClassName changes.

Fix for https://github.com/tractorcow-farm/silverstripe-fluent/issues/713

### CODE REVIEW

**Comment by mfendeksilverstripe** [comment]
@tractorcow @chrispenny Please review and let me know your feedback.

File: src/Extension/FluentExtension.php [hunk 1]
```diff
@@ -593,16 +594,18 @@ public function augmentSQL(SQLSelect $query, DataQuery $dataQuery = null)
...
-
-        /** @var string $currentLocale */
         $currentLocale = FluentState::singleton()->getLocale();
+
         if (!$currentLocale) {
             return;
         }
 
+        // Detect class name change (used later to properly localised newly available relation data)
+        $owner->classNameChangeDetected = $owner->isInDB() && $owner->isChanged('ClassName');
```
Comment (tractorcow) [inline src/Extension/FluentExtension.php:607]: you can detect isChanged in onAfterWrite. I prefer to do this there than adding an unnecessary temp variable.

You can also use isChanged('ID') to see if a record was created.

File: src/Extension/FluentExtension.php [hunk 2]
```diff
@@ -617,6 +620,114 @@ public function onBeforeWrite()
...
+                    }
+                );
+
+                if (!$localisedRecord) {
+                    // This is just a sanity check
+                    continue;
+                }
+
+                $relationIdField = $relation . 'ID';
+                $localisedRelationID = $localisedRecord->{$relationIdField};
```
Comment (tractorcow) [inline src/Extension/FluentExtension.php:696]: what if relation is polymorphic? (ID + ClassName field both).

File: src/Extension/FluentExtension.php [hunk 3]
```diff
@@ -617,6 +620,114 @@ public function onBeforeWrite()
...
+                // Attach the duplicated relation to localised record
+                FluentState::singleton()->withState(
+                    function (FluentState $state) use ($locale, $localisedRecord, $relation, $duplicate): void {
+                        $state->setLocale($locale->Locale);
+
+                        $localisedRecord->setComponent($relation, $duplicate);
+
+                        if ($localisedRecord->hasExtension(Versioned::class)) {
+                            /** @var DataObject|Versioned $localisedRecord */
+                            $localisedRecord->writeWithoutVersion();
```
Comment (tractorcow) [inline src/Extension/FluentExtension.php:719]: Why write without versioned? Just a ->write() in all cases would be enough?

File: src/Extension/FluentExtension.php [hunk 4]
```diff
@@ -617,6 +620,114 @@ public function onBeforeWrite()
...
+        if ($stage !== Versioned::DRAFT) {
+            // Only draft stage is relevant for the duplication
+            return;
+        }
+
+        // Get list of all localised instances of this model and duplicate relations if needed (if current one has it)
+        $locales = $owner->getLocaleInstances();
+
+        /** @var Locale $locale */
+        foreach ($locales as $locale) {
```
Comment (tractorcow) [inline src/Extension/FluentExtension.php:659]: maybe if you loop over $this->Locales() instead, you can re-use the logic in RecordLocale to simplify some of this code.

E.g. 


```php
$localisedRecord = $recordLocale->getRecord();
```

**Review by tractorcow — CHANGES_REQUESTED** [review]

**Comment by tractorcow** [comment]
@mfendeksilverstripe looks good; Asked some questions, and added some suggestions that can simplify the code a bit and re-use existing work :)

File: src/Extension/FluentExtension.php [hunk 5]
```diff
@@ -617,6 +620,114 @@ public function onBeforeWrite()
...
+            if ($locale->Locale === $currentLocale) {
+                // We only need to handle other locale instances, current locale doesn't need updates
+                continue;
+            }
+
+            $relations = (array) $owner->config()->get('localised_copy');
```
Comment (tractorcow) [inline src/Extension/FluentExtension.php:665]: localised_copy is exclusively for has_one right? just checking :)

Comment (mfendeksilverstripe) [reply src/Extension/FluentExtension.php:665]: That's correct.

File: src/Extension/FluentExtension.php (hunk 4 above)
Comment (mfendeksilverstripe) [reply src/Extension/FluentExtension.php:659]: I had a look at this but it I couldn't find a way to simplify this further - the goal is to iterate over all locales which have localised record except the current locale. I'm already using the existing functionality to search for such locales via `getLocaleInstances`. I can simplify how the data changes are made though (one write per model).

File: src/Extension/FluentExtension.php (hunk 2 above)
Comment (mfendeksilverstripe) [reply src/Extension/FluentExtension.php:696]: I think it's sufficient to check only ID for the purpose of validating if we need to do an update or not. Note that later, when we actually assign the new relation we use `setComponent` which also covers the ClassName.

File: src/Extension/FluentExtension.php (hunk 4 above)
Comment (tractorcow) [reply src/Extension/FluentExtension.php:659]: you could put those into a condition and `continue` in that case. still less code, and you can use RecordLocale to check if the localisation exists.

Comment (mfendeksilverstripe) [reply src/Extension/FluentExtension.php:659]: Not sure I follow - we are already looping a list of locales in which the model is localised (has localised record). I'm only skipping the current locale as I can't exclude it from the list as `getLocaleInstances` returns an `array` not an `ArrayList`.

**Comment by mfendeksilverstripe** [comment]
@tractorcow I've pushed up new changes, please review.

File: src/Extension/FluentExtension.php (hunk 4 above)
Comment (tractorcow) [reply src/Extension/FluentExtension.php:657]: I think the problem I have is that getLocaleInstances() is solving the same problem that the Locales() method had already solved, so we end up with a lot of duplication of effort.

without this method, we could do this


```
foreach ($this->Locales() as $recordLocale) {
  // simpler "saved in this locale?" check
  if (!$recordLocale->IsDraft()) {
    continue;
  } 
$localisedRecord  = $recordLocale->getRecord();
// rest of code
}
```

File: src/Extension/FluentExtension.php (hunk 2 above)
Comment (tractorcow) [reply src/Extension/FluentExtension.php:696]: the ID of two different polymorphic classes can be the same, but those are different objects. :)

E.g. polymorphic relation to a File ID = 5 could be replaced with a Page ID = 5.

**Comment by tractorcow** [comment]
Hi @mfendeksilverstripe want to say again what a great job you have done on this PR so far. I am being pedantic I know, but I want to check I have covered every base to save issues later on down the line. :)

Let me know if you want me to do a suggested patch to this PR before we merge; I want to do a semi-review of the codebase before 6.0.0 anyway.

**Comment by tractorcow** [comment]
hi @mfendeksilverstripe I had more of a think about it, and I think we can probably worry about optimisation / code cleanup later. I have to spend a bit of time to do a review of master before releasing 6.0.0 anyway.

The only fix I think is blocking is the polymorphic relation, which is a real bug still.

File: src/Extension/FluentExtension.php (hunk 4 above)
Comment (tractorcow) [reply src/Extension/FluentExtension.php:657]: As mentioned below, I can tidy this up later; better to get what's already done in the code for now. :)

File: src/Extension/FluentExtension.php (hunk 2 above)
Comment (tractorcow) [reply src/Extension/FluentExtension.php:696]: you could just query the objects directly without trying to check the fields

```php
$ownerRelation = $owner->{$relation}();
$localisedRelation = $localisedRecord->{$relation}();
if ($ownerRelation->ID !== $localisedRelation->ID || get_class($ownerRelation) !== get_class($localisedRelation)) {
  continue;
}
```

**Comment by tractorcow** [comment]
I left a suggestion above on a quick and easy way to solve that :D

**Comment by mfendeksilverstripe** [comment]
Thanks for help @tractorcow , I added the small changes you requested. Please review.

**Review by tractorcow — APPROVED** [review]

**Comment by tractorcow** [comment]
it's done :D

**Comment by tractorcow** [comment]
I made my suggested cleanup changes in https://github.com/tractorcow-farm/silverstripe-fluent/pull/719

**Pull request merged**